ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Restore.hs

```diff
@@ -0,0 +1,181 @@
+-- | Cardano Wallet restoration logic.
+
+module Ariadne.Wallet.Cardano.Kernel.Restore
+       ( collectUtxo
+       , restoreWallets
+       , RestoreFrom (..)
+       , WalletToRestore(..)
+       ) where
+
+import qualified Universum.Unsafe as Unsafe (init)
+
+import Control.Exception (Exception(displayException))
+import Control.Lens (at, non, (?~))
+import qualified Data.ByteString as BS
+import qualified Data.Map.Strict as Map
+
+import Pos.Binary.Class (decodeFull')
+import Pos.Core.Configuration (HasConfiguration)
+import Pos.Crypto (EncryptedSecretKey, PassPhrase)
+import qualified Pos.Crypto as Crypto
+  (checkPassMatches, deriveHDPassphrase, emptyPassphrase, encToPublic,
+  safeDeterministicKeyGen)
+import Pos.Txp.Toil.Types (Utxo)
+import Pos.Util.BackupPhrase (BackupPhrase(..), safeKeysFromPhrase)
+import Pos.Util.UserSecret (usKeys0)
+
+import Ariadne.Cardano.Face (Address, CardanoMode)
+import Ariadne.Wallet.Cardano.Kernel.AddressDiscovery
+  (AddressWithPathToUtxoMap, discoverHDAddressWithUtxo)
+import Ariadne.Wallet.Cardano.Kernel.Bip32 (DerivationPath(..))
+import Ariadne.Wallet.Cardano.Kernel.Bip39 (mnemonicToSeedNoPassword)
+import Ariadne.Wallet.Cardano.Kernel.Bip44
+  (Bip44DerivationPath(..), bip44PathToAddressId, decodeBip44DerivationPath)
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
+  (AssuranceLevel(..), HdAddressId, WalletName(..), eskToHdRootId,
+  hdAddressIdParent)
+import Ariadne.Wallet.Cardano.Kernel.PrefilterTx
+  (PrefilteredUtxo, UtxoByAccount)
+import Ariadne.Wallet.Cardano.Kernel.Wallets
+  (HasNonemptyPassphrase(..), mkHasPP)
+import Ariadne.Wallet.Face (Mnemonic(..))
+
+newtype WrongMnemonic = WrongMnemonic Text
+ deriving (Eq, Show)
+
+instance Exception WrongMnemonic where
+  displayException (WrongMnemonic txt) =
+    "Wrong mnemonic: " <> show txt
+
+data SecretsDecodingError = SecretsDecodingError FilePath Text
+  deriving (Eq, Show)
+
+instance Exception SecretsDecodingError where
+  displayException (SecretsDecodingError path txt) =
+    "Failed to decode " <> path <> ": " <> show txt
+
+data RestoreFrom
+  = RestoreFromMnemonic !Mnemonic
+                        !PassPhrase
+  | RestoreFromKeyFile  !FilePath
+
+
+data WalletToRestore
+  = WalletToRestore
```

Comment:
I believe that our code style says that constructor should be on the previous line in this case.